cups/ppd-cache.c

```diff
@@ -3435,9 +3435,10 @@ _ppdCreateFromIPP2(
 
   if ((attr = ippFindAttribute(supported, "document-format-supported", IPP_TAG_MIMETYPE)) != NULL)
   {
-    is_apple = ippContainsString(attr, "image/urf");
+    is_apple = ippContainsString(attr, "image/urf") && (ippFindAttribute(supported, "urf-supported", IPP_TAG_KEYWORD) != NULL);
     is_pdf   = ippContainsString(attr, "application/pdf");
-    is_pwg   = ippContainsString(attr, "image/pwg-raster") && !is_apple;
+    is_pwg   = ippContainsString(attr, "image/pwg-raster") && !is_apple &&
+	       (ippFindAttribute(supported, "pwg-raster-document-resolution-supported", IPP_TAG_KEYWORD) != NULL);
```

Comment:
You should also look for "pwg-raster-document-type-supported"...

Reply:
I thought about the attribute as well (since it is required in the standard too), but according this code further down:

```
4156   if ((attr = ippFindAttribute(supported, "urf-supported", IPP_TAG_KEYWORD)) == NULL)
4157     if ((attr = ippFindAttribute(supported, "pwg-raster-document-type-supported", IPP_TAG_KEYWORD)) == NULL)
4158       if ((attr = ippFindAttribute(supported, "print-color-mode-supported", IPP_TAG_KEYWORD)) == NULL)
4159         attr = ippFindAttribute(supported, "output-mode-supported", IPP_TAG_KEYWORD);
```

we fallback to print-color-mode-supported and output-mode-supported if the attribute is missing. We don't use the attribute anywhere else (AFAIK) in the function, so requiring the attribute would break the PPD generation for such devices.
I aimed PR to match with the existing code, so that's why I didn't check the attribute.

I can add the check, but then I would make further changes. So I see the following possibilities:

1. add check for the mentioned attribute, and rewrote the code snippet I mentioned to don't do fallback if we have PWG raster,
2. add extensive check in is_pwg line, incorporating fallbacks there as well,
3. has the PR as it is.

IMO option 3 provides wider support for possibly broken devices, so I would go with that.

Reply:
@zdohnal Technically the lack of urf-supported or pwg-raster-document-type-supported means that we have a very low chance of anything working (i.e. if the printer supports PDF but isn't able to print the PDF file we send it, what kind of raster will it accept if it doesn't tell us?)

urf-supported combines the three pwg-raster-document-xxx attributes - resolution (RSn-n-n), type (Wnn, SRGBnn, ADOBERGBnn, etc.), and sheet back (DMn) - that are needed for properly generating raster data. So the Apple Raster (URF) check needs to look at one attribute while the PWG Raster check needs to look at more (pwg-raster-document-{resolution-supported,sheet-back,type-supported} are used for resolution, duplex printing, and color space/bit depth).

Reply:
@michaelrsweet ok, makes sense this way. I'll add check for pwg-raster-document-type-supported.

@tillkamppeter do you want to add it to libppd as well (through direct push to the repo)?

Reply:
@zdohnal OK, will do for libppd.

Reply:
Done in OpenPrinting/libppd@c7a62e8c4c37. Thanks for the hint @michaelrsweet and @zdohnal !